servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/apple/traffic/resilience/http/AbstractTrafficManagementHttpFilter.java

```diff
@@ -0,0 +1,373 @@
+        return offloadNone();
+    }
+
+    // Each filter needs a new capacity & breaker partitions functions.
+    final Function<HttpRequestMetaData, CapacityLimiter> newCapacityPartitions() {
+        return capacityPartitionsSupplier.get();
+    }
+
+    final Function<HttpRequestMetaData, CircuitBreaker> newCircuitBreakerPartitions() {
+        return circuitBreakerPartitionsSupplier.get();
+    }
+
+    Single<StreamingHttpResponse> applyCapacityControl(
+            final Function<HttpRequestMetaData, CapacityLimiter> capacityPartitions,
+            final Function<HttpRequestMetaData, CircuitBreaker> circuitBreakerPartitions,
+            @Nullable final ServerListenContext serverListenContext,
+            final StreamingHttpRequest request,
+            @Nullable final StreamingHttpResponseFactory responseFactory,
+            final Function<StreamingHttpRequest, Single<StreamingHttpResponse>> delegate) {
+        return defer(() -> {
+            final long startTime = nanoTime();
+            final CapacityLimiter partition = capacityPartitions.apply(request);
+            if (partition == null) {
+                observer.onRejectedUnmatchedPartition(request);
+                return rejectWhenNotMatchedCapacityPartition ?
+                        handleLocalCapacityRejection(null, request, responseFactory)
+                                .shareContextOnSubscribe() :
+                        handlePassthrough(delegate, request)
+                                .shareContextOnSubscribe();
+            }
+
+            final CircuitBreaker breaker = circuitBreakerPartitions.apply(request);
+            final ContextMap meta = request.context();
+            final Classification classification = classifier.apply(request);
+            Ticket ticket = partition.tryAcquire(classification, meta);
+
+            if (ticket != null) {
+                ticket = new TrackingDelegatingTicket(ticket, request.hashCode());
+            }
+
+            if (ticket == null) {
+                observer.onRejectedLimit(request, partition.name(), meta, classification);
+                return handleLocalCapacityRejection(serverListenContext, request, responseFactory)
+                        .shareContextOnSubscribe();
+            } else if (breaker != null && !breaker.tryAcquirePermit()) {
+                observer.onRejectedOpenCircuit(request, breaker.name(), meta, classification);
+                // Ignore the acquired ticket if breaker was open.
+                ticket.ignored();
+                return handleLocalBreakerRejection(request, responseFactory, breaker).shareContextOnSubscribe();
+            }
+
+            // Ticket lifetime must be completed at all points now, try/catch to ensure if anything throws (e.g.
+            // reactive flow isn't followed) we still complete ticket lifetime.
+            try {
+                final TicketObserver ticketObserver = observer.onAllowedThrough(request, ticket.state());
```

Comment:
The `ticket` implementation could through, depending on how a user implements it.

Reply:
But `ticket` is not invoked inside `handleAllow` directly, it's invoked inside a reactive chain build by `handleAllow`.

Reply:
The ticket is involved in the line above
```
final TicketObserver ticketObserver = observer.onAllowedThrough(request, ticket.state());
```
it feels better to define the observer with then try-catch block and use it instantly, rather than have the try-catch modify a variable out-of-scope, to use it later.

Reply:
you're right, overlooked that one, thanks!